### PR TITLE
chain: Refund transaction fee if success with all report messages

### PR DESCRIPTION
**Pull request by sorawit**

### CODE REVIEW

**Review by prin-r — COMMENTED** [review]
This pr was tested by comparing report tx with send tx.
Both txs were sent by validator1 with fee = 500,000 uband.
Validator1 lose nothing after sending report tx but lose 500,888 uband for sending tx (fee 500,00, amount 888).

[reportTx.txt](https://github.com/bandprotocol/bandchain/files/4338285/reportTx.txt)
[sendTx.txt](https://github.com/bandprotocol/bandchain/files/4338287/sendTx.txt)

File: chain/app/app.go [hunk 1]
```diff
@@ -361,20 +343,56 @@ func NewBandApp(
...
+					isAllReportTxs = false
+				}
+			}
+			if isAllReportTxs {
+				acc := app.accountKeeper.GetAccount(app.deliverContext, stdTx.GetSigners()[0])
+				if acc != nil {
+					err := app.supplyKeeper.SendCoinsFromModuleToAccount(
+						app.deliverContext,
+						auth.FeeCollectorName,
+						acc.GetAddress(),
```
Comment (taobun) [inline chain/app/app.go]: Why do we need to get an account from the account keeper? Using stdTx.GetSigners()[0] as to address from send coin is not enough?

File: chain/app/app.go [hunk 2]
```diff
@@ -361,20 +343,56 @@ func NewBandApp(
...
+						stdTx.Fee.Amount,
```
Comment (taobun) [inline chain/app/app.go]: wow feels good

File: chain/app/app.go (hunk 1 above)
Comment (sorawit) [reply chain/app/app.go]: I'm not sure, I just strictly follow what Cosmos ante code is doing. Can you help double check? Cosmos `x/auth/ante.go` specifically calls `GetSignerAcc` to get account from address, then call `DeductFees` on the account, though `DeductFees` just calls `SendCoinsFromModuleToAccount` on the acccount's address.

Comment (sorawit) [reply chain/app/app.go]: Is it possible that the account is removed somehow and somehow reach here without an account on that address?

File: chain/app/app.go [hunk 3]
```diff
@@ -123,6 +122,11 @@ type bandApp struct {
 	paramsKeeper   params.Keeper
 	zoracleKeeper  zoracle.Keeper
 
+	// Decoder for unmarshaling []byte into sdk.Tx
+	txDecoder sdk.TxDecoder
+	// Deliver Context that is set during BeingBlock and unset during EndBlock; primarily for gas refund
+	deliverContext sdk.Context
```
Comment (taobun) [inline chain/app/app.go:128]: Make it to public or create getter function to get it

File: chain/app/app.go [hunk 4]
```diff
@@ -361,20 +343,54 @@ func NewBandApp(
...
+		// Refund 100% of gas fee for any successful transaction that only contains MsgReportData
+		tx, err := app.txDecoder(req.Tx)
+		if err != nil { // Should never happen because BaseApp.DeliverTx succeeds
+			panic(err)
+		}
+		isAllReportTxs := true
+		if stdTx, ok := tx.(auth.StdTx); ok {
+			for _, msg := range tx.GetMsgs() {
+				if _, ok := msg.(zoracle.MsgReportData); !ok {
+					isAllReportTxs = false
```
Comment (taobun) [inline chain/app/app.go:373]: break loop?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/app.go (hunk 3 above)
Comment (sorawit) [reply chain/app/app.go:128]: That will be on a separate PR to make everything including all keepers accessible.

File: chain/app/app.go (hunk 4 above)
Comment (sorawit) [reply chain/app/app.go:373]: Right. thanks!

**Review by taobun — APPROVED** [review]

**Pull request merged**